File: Maximum_depth_of_binary_tree/solution_postorderTraversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return an integer
    def maxDepth(self, root):
        stack = []
        visited = None
        cur = root
        max_depth = 0
        while cur is not None or len(stack)!=0:
            while cur is not None:
                stack.append(cur)
                cur = cur.left
            cur = stack.pop()
            stack.append(cur)
            if cur.left is None and cur.right is None:
                if max_depth<(len(stack)):
                    max_depth = len(stack)
            if cur.right is not None and cur.right is not visited:
                cur = cur.right
            else:
                stack.pop()
                visited = cur
                cur = None
        return max_depth
            
        

File: Maximum_depth_of_binary_tree/solution_postorderTraversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode *root) {
        int max_depth = 0;
        stack<TreeNode*> s;
        TreeNode *cur,*visited;
        cur = root;
        while(cur || !s.empty()){
            while(cur){
                s.push(cur);
                cur = cur->left;
            }
            cur = s.top();
            //for every leafnode, the stack has its path to root
            if(!cur->left && !cur->right){
                if(s.size()>max_depth){
                    max_depth = s.size();
                }
            }
            if(cur->right && cur->right!=visited){
                cur = cur->right;
            }else{
                s.pop();
                visited = cur;
                cur = NULL;
            }
        }
        return max_depth;
    }
};